src/librustc_mir/transform/generator.rs

```diff
@@ -1072,14 +1072,27 @@ fn create_generator_resume_function<'tcx>(
     use rustc::mir::interpret::PanicInfo::{
         GeneratorResumedAfterPanic,
         GeneratorResumedAfterReturn,
+        AsyncResumedAfterReturn,
+        AsyncResumedAfterPanic,
     };
 
     // Jump to the entry point on the unresumed
     cases.insert(0, (UNRESUMED, BasicBlock::new(0)));
-    // Panic when resumed on the returned state
-    cases.insert(1, (RETURNED, insert_panic_block(tcx, body, GeneratorResumedAfterReturn)));
-    // Panic when resumed on the poisoned state
-    cases.insert(2, (POISONED, insert_panic_block(tcx, body, GeneratorResumedAfterPanic)));
+
+    // Panic when resumed on the returned or poisoned state
+    match body.generator_kind {
+        Some(GeneratorKind::Async(_)) => {
+            cases.insert(1, (RETURNED, insert_panic_block(tcx, body, AsyncResumedAfterReturn)));
+            cases.insert(2, (POISONED, insert_panic_block(tcx, body, AsyncResumedAfterPanic)));
+        },
+        Some(GeneratorKind::Gen) => {
+            cases.insert(1, (RETURNED, insert_panic_block(tcx, body, GeneratorResumedAfterReturn)));
+            cases.insert(2, (POISONED, insert_panic_block(tcx, body, GeneratorResumedAfterPanic)));
+        },
+        None => {
+            // N/A because we would never create a resume function if there was no generator_kind
+        }
+    };
```

Comment:
Maybe less duplication if it's expressed as:
```
let (after_return, after_panic) = match body.generator_kind { 
   Some(GeneratorKind::Async(_)) => (AsyncResumedAfterReturn, AsyncResumedAfterPanic),
   Some(GeneratorKind::Gen) => (GeneratorResumedAfterReturn, GeneratorResumedAfterPanic)
}
cases.insert(1, (RETURNED, insert_panic_block(tcx, body, after_return)));
cases.insert(2, (POISONED, insert_panic_block(tcx, body, after_panic)));
```

Reply:
Good idea, I'll do that. Thanks!